### PR TITLE
Add support for message components

### PR DESCRIPTION
**Pull request by ire4ever1190**
Adds support for all current message components (ActionRow, SelectMenu, Button) and added an example to go with it

Found working with the objects directly was a bit annoying and added a simple api to create the components easy enough (this can easily be removed if needed)

Tested and it works correctly (might be some edge cases I missed so if anyone wants to toy around and find them, feel free lol)

### CODE REVIEW

**Comment by dannyhpy** [comment]
Hey!

Discord also supports **Message components** for webhooks
https://discord.com/developers/docs/resources/webhook#execute-webhook

**Comment by ire4ever1190** [comment]
Thanks! Will add that in asap

**Comment by krisppurg** [comment]
Thanks hobble for the contribution.

As of recently I have a bunch of files unsaved (because the implementation is uncomplete.).
Hopefully I will get to save those files, test them and finally push them to master.

I will review your Pull Request and of course change what is needed 👍.

**Comment by ire4ever1190** [comment]
Added components for `executeWebhook` and merged upstream

**Comment by krisppurg** [comment]
Nice

**Comment by krisppurg** [comment]
I will go back to my left over files and complete some of them. Hopefully it doesn't take long.

File: dimscord/objects.nim [hunk 1]
```diff
@@ -879,3 +879,137 @@ proc newApplicationCommand*(data: JsonNode): ApplicationCommand =
...
+
+proc len*(component: MessageComponent): int =
+    ## Returns number of items in an ActionRow or number of options in a menu
+    case component.kind:
+        of ActionRow:
+            result = component.components.len
+        of SelectMenu:
+            result = component.options.len
+        else:
+            raise newException(ValueError, "Component must be ActionRow or SelectMenu")
```
[the remark this answers is not in the document]
Comment (ire4ever1190) [reply dimscord/objects.nim]: Done, moved helpers into `helper.nim`

File: dimscord/objects.nim [hunk 2]
```diff
@@ -879,3 +879,137 @@ proc newApplicationCommand*(data: JsonNode): ApplicationCommand =
...
+    # Beware, this check might be invalid in future when more
+    # components are added
+    assert contains.len <= 1, "Action rows can only contain one type"
+    if contains.hasKey(SelectMenu):
+        assert contains[SelectMenu] == 1, "Can only have one select menu per action row"
+        assert row.components[0].options.len > 0, "Menu must have options"
+    elif contains.hasKey(Button):
+        assert contains[Button] <= 5, "Can only have <= 5 buttons per row"
+    else:
+        assert not contains.hasKey(ActionRow), "Action row cannot contain an action row"
```
[the remark this answers is not in the document]
Comment (ire4ever1190) [reply dimscord/objects.nim]: Same as above

**Comment by ire4ever1190** [comment]
There seems to be some issues with your latest commit, I'll try and solve it before this gets merged

**Comment by krisppurg** [comment]
Is it fixed now?

**Comment by ire4ever1190** [comment]
Sorry for the delay, I've ran into emojis that I didn't have before

**Comment by ire4ever1190** [comment]
Seems to all be working again

**Review by krisppurg — APPROVED** [review]

**Pull request merged**